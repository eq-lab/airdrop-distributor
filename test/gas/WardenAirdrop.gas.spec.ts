import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import snapshotGasCost from '@uniswap/snapshot-gas-cost';
import { ethers } from 'hardhat';
import { createAirdrop } from '../shared/fixtures';
import { constructMerkleTree, generateRandomAirdropData } from '../shared/utils';
import { parseUnits } from 'ethers/lib/utils';

describe('WardenAirdrop gas', () => {
  it('claim gas', async () => {
    const { wardenAirdrop, owner } = await loadFixture(createAirdrop);
    const [_, user] = await ethers.getSigners();

    const airdropData = generateRandomAirdropData(99);
    const userAmount = parseUnits('100', 18);
    airdropData.push({ address: user.address, amount: userAmount });

    const tree = constructMerkleTree(airdropData);
    await wardenAirdrop.connect(owner).updateRoot(tree.getHexRoot());

    const userProof = tree.getHexProof(
      ethers.utils.solidityKeccak256(['address', 'uint256'], [user.address, userAmount])
    );
    await snapshotGasCost(await wardenAirdrop.connect(user).claim(userAmount, userProof));
  });

  it('updateRoot gas 100 users', async () => {
    const { wardenAirdrop, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(100);
    const tree = constructMerkleTree(airdropData);
    await snapshotGasCost(await wardenAirdrop.connect(owner).updateRoot(tree.getHexRoot()));
  });

  it('updateRoot gas 1000 users', async () => {
    const { wardenAirdrop, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(1000);
    const tree = constructMerkleTree(airdropData);
    await snapshotGasCost(await wardenAirdrop.connect(owner).updateRoot(tree.getHexRoot()));
  });

  it('updateRoot gas 10000 users', async () => {
    const { wardenAirdrop, owner } = await loadFixture(createAirdrop);

    const airdropData = generateRandomAirdropData(10000);
    const tree = constructMerkleTree(airdropData);
    await snapshotGasCost(await wardenAirdrop.connect(owner).updateRoot(tree.getHexRoot()));
  });
});
