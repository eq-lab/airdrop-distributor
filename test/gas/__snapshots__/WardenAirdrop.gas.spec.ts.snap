// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WardenAirdrop gas claim gas 100 eligible users 1`] = `
Object {
  "calldataByteLength": 324,
  "gasUsed": 95480,
}
`;

exports[`WardenAirdrop gas claim gas 1000 eligible users 1`] = `
Object {
  "calldataByteLength": 420,
  "gasUsed": 98445,
}
`;

exports[`WardenAirdrop gas claim gas 10000 eligible users 1`] = `
Object {
  "calldataByteLength": 548,
  "gasUsed": 102207,
}
`;

exports[`WardenAirdrop gas claim gas 50000 eligible users 1`] = `
Object {
  "calldataByteLength": 612,
  "gasUsed": 103697,
}
`;

exports[`WardenAirdrop gas updateRoot gas 100 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;

exports[`WardenAirdrop gas updateRoot gas 1000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;

exports[`WardenAirdrop gas updateRoot gas 10000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;

exports[`WardenAirdrop gas updateRoot gas 50000 users 1`] = `
Object {
  "calldataByteLength": 36,
  "gasUsed": 47141,
}
`;
